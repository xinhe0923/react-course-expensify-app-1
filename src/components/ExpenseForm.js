import React from "react";
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

//const date =new Date();
// const now = moment();
// console.log(now.format("MMM Do YYYY"));

export default class ExpenseForm extends React.Component {
  constructor(props){
    super(props);
    this.state= {
      description: props.expense? props.expense.description : '',
      note: props.expense? props.expense.note: "",
      amount: props.expense? (props.expense.amount/100).toString(): "",
      createdAt: props.expense? moment(props.expense.createdAt): moment(), //this is the initial value for the calender day picker
      calenderFocused: false,
      error: "",
    };

  }
 
  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };
  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  };
  onAmountChange = (e) => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({
        amount,
      }));
    }
  };
  onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({
        createdAt,
      }));
    }
  };
  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calenderFocused: focused }));
  };

  onSubmit = (e) => {
    e.preventDefault(); //no full page refresh
    if (!this.state.description || !this.state.amount) {
      //set error state eqaul to 'please provide description and amount'
      this.setState(() => ({
        error: "please provide description and amount",
      }));
    } else {
      //clear the error
      this.setState(() => ({
        error: "",
      }));
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10)*100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note,
      });
    }
  };
  render() {
    return (
      <div>
        {this.state.error && <p>{this.state.error}</p>}
        <form onSubmit={this.onSubmit}>
          <input
            type="text"
            placeholder="Description"
            autoFocus
            value={this.state.description}
            onChange={this.onDescriptionChange} //change the value?
          />
          <input
            type="text"
            placeholder="Amount"
            value={this.state.amount}
            onChange={this.onAmountChange}
          />
          <SingleDatePicker
            date={this.state.createdAt} //the moment object represents when you wanna start
            onDateChange={this.onDateChange}
            focused={this.state.calenderFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <textarea
            type="text"
            placeholder="add a note for your expense"
            value={this.state.note}
            onChange={this.onNoteChange}
          ></textarea>
          <button>add expenses</button>
        </form>
      </div>
    );
  }
}
