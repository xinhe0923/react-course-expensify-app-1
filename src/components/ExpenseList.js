import React from "react";
import { connect } from "react-redux";
import ExpenseListItem from "./ExpenseListItem";
import selectExpenses from "../selectors/expenses";

export const ExpenseList = (props) => (
  <div className="content-container">
    <div className="list-header">
      <div className="show-for-mobile">Expenses</div>
      <div className="show-for-desktop">Expense</div>
      <div className="show-for-desktop">Amount</div>
    </div>
    <div className='list-body'>
    {props.expenses.length === 0 ? (
      <div className='list-item list-item--message'>
        <span>no expenses</span>
      </div>
    ) : (
      props.expenses.map((item) => {
        return <ExpenseListItem key={item.id} {...item} />;
      })
    )}
    </div>
  </div>
);

const mapStateToProps = (state) => {
  return {
    expenses: selectExpenses(state.expenses, state.filters), //the expenselist component now has access to the name
  };
}; //get the function back
//the first argument is what kind of information we wanna connect

export default connect(mapStateToProps)(ExpenseList);
