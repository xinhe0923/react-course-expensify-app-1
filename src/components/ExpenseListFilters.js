import React from "react";
import { connect } from "react-redux";
import {
  setTextFilter,
  sortByDate,
  sortByAmount,
  setStartDate,
  setEndDate,
} from "../actions/filters";
import { DateRangePicker } from "react-dates";

export class ExpenseListFilters extends React.Component {
  state = {
    calenderFocused: null,
  };
  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };
  onFocusChange = (calenderFocused) => {
    this.setState(() => ({
      calenderFocused,
    }));
  };
  onTextChange = (e) => {
    this.props.setTextFilter(e.target.value);
  };
  onSortChange = (e) => {
    if (e.target.value === "date") {
      this.props.sortByDate();
    } else if (e.target.value === "amount") {
      this.props.sortByAmount();
    }
  };
  render() {
    return (
      <div className="content-container">
        <div className="input-group">
          <div className="input-group__item">
            <input
            placeholder="search expenses"
              type="text"
              className="text-input"
              value={this.props.filters.text}
              onChange={this.onTextChange}
            />
          </div>
          <div className="input-group__item">
            <select
            className='select'
              value={this.props.filters.sortBy} //as the store changes, we need to set the value
              //this is controlled input.
              onChange={this.onSortChange}
            >
              <option value="date">Date</option>
              <option value="amount">Amount</option>
            </select>
          </div>
          <div className="input-group__item">
            <DateRangePicker
              startDate={this.props.filters.startDate}
              endDate={this.props.filters.endDate}
              onDatesChange={this.onDatesChange}
              focusedInput={this.state.calenderFocused}
              onFocusChange={this.onFocusChange}
              numberOfMonths={1}
              isOutsideRange={() => false}
              showClearDates={true}
            />
          </div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state) => ({
  filters: state.filters,
});

const mapDispatchToProps = (dispatch) => ({
  setTextFilter: (text) => dispatch(setTextFilter(text)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount()),
  setStartDate: (startDate) => dispatch(setStartDate(startDate)),
  setEndDate: (endDate) => dispatch(setEndDate(endDate)),
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);
