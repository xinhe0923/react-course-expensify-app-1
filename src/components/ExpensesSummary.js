import React from "react";
import { connect } from "react-redux";
import { Link } from "react-router-dom";
import selectExpensesTotal from "../selectors/expenses-total";
import selectExpenses from "../selectors/expenses";
import numeral from "numeral";

export const ExpensesSummary = ({ expenseCount, expenseTotal }) => {
  const expenseWord = expenseCount === 1 ? "expense" : "expenses";
  const formatedExpensesTotal = numeral(expenseTotal / 100).format("$0,0.00");
  return (
    <div className="page-header">
      <div className="content-container">
        <h1 className="page-header__title">
          viewing <span>{expenseCount} </span>
          {expenseWord} totaling <span>{formatedExpensesTotal}</span>
        </h1>
        <div className="page-header__actions">
          <Link className="button" to="/create">
            Add Expense
          </Link>
        </div>
      </div>
    </div>
  );
};

const mapStateToProps = (state) => {
  const visibleExpenses = selectExpenses(state.expenses, state.filters);
  return {
    expenseCount: visibleExpenses.length,
    expenseTotal: selectExpensesTotal(visibleExpenses), //the expenselist component now has access to the name
  };
}; //get the function back
//the first argument is what kind of information we wanna connect

export default connect(mapStateToProps)(ExpensesSummary);
