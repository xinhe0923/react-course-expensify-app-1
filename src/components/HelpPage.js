import React from "react";

const HelpPage=()=>(
    <div>this is from my HelpPage component</div>

)

export default HelpPage;