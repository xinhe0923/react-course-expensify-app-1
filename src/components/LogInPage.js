import React from "react";
import { connect } from "react-redux";
import { startLogin } from "../actions/auth";


export const LogInPage = ({startLogin}) => (//destrructure
    <div className="box-layout" >
      <div className='box-layout__box'>
        <h1 className="box-layout_title">Expensify</h1>
      <p>It's time to get your expenses under control</p>
      <button className="button" onClick={startLogin} >Log in with Google</button>

      </div>

    </div>
  );

const mapDispatchToProps=(dispatch)=>({
startLogin:()=>dispatch(startLogin())
}  )
export default connect(undefined,mapDispatchToProps)(LogInPage)