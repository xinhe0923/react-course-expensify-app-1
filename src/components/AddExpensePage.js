import React from "react";
import { connect } from "react-redux";
import ExpenseForm from "./ExpenseForm";
import { addExpense } from "../actions/expenses";

export class AddExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.addExpense(expense);
    this.props.history.push("/");
  };
  render() {
    return (
      <div>
        <h1> AddExpensePage </h1>
        <ExpenseForm onSubmit={this.onSubmit} />
      </div>
    );
  }
}

// const AddExpensePage = (props) => (
//   <div>
//     <h1> AddExpensePage </h1>
//     <ExpenseForm
//       onSubmit={(expense) => {
//         // props.dispatch(addExpense(expense));
//         props.onSubmit(expense);
//         props.history.push("/");
//       }}
//     />
//   </div>
// );

const mapDispatchToProps = (dispatch) => ({
  addExpense: (expense) => dispatch(addExpense(expense)),
});

//extract the expense data from expenseForm because we need to dispatch
//this part of data in 2 files:AddExpensePage,EditExpensePage

export default connect(undefined, mapDispatchToProps)(AddExpensePage);
