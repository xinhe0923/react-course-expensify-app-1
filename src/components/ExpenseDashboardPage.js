import React from "react";
import ExpenseList from "./ExpenseList";
import ExpenseListFilters from "./ExpenseListFilters";
import  ExpensesSummary  from "./ExpensesSummary";
import Header from "./Header";

const ExpenseDashboardPage = () => (
  <div>
    <ExpenseListFilters/>

    <ExpenseList />
    <ExpensesSummary/>
  </div>
);
export default ExpenseDashboardPage;
