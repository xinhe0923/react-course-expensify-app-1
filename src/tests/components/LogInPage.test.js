import React from "react";
import { shallow } from "enzyme";
import { LogInPage } from "../../components/LogInPage";

test("should render LogInPage", () => {
  const wrapper = shallow(<LogInPage />);
  expect(wrapper).toMatchSnapshot();
});
test("should call startLogin on button lick", () => {
  const startLogin = jest.fn();

  const wrapper = shallow(<LogInPage startLogin={startLogin} />);
  expect(wrapper).toMatchSnapshot();
  wrapper.find("button").simulate("click");

  expect(startLogin).toHaveBeenCalled();
});
