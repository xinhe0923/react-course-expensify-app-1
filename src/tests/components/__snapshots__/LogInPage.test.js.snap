// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should call startLogin on button lick 1`] = `
<div
  className="box-layout"
>
  <div
    className="box-layout__box"
  >
    <h1
      className="box-layout_title"
    >
      Expensify
    </h1>
    <p>
      It's time to get your expenses under control
    </p>
    <button
      className="button"
      onClick={[Function]}
    >
      Log in with Google
    </button>
  </div>
</div>
`;

exports[`should render LogInPage 1`] = `
<div
  className="box-layout"
>
  <div
    className="box-layout__box"
  >
    <h1
      className="box-layout_title"
    >
      Expensify
    </h1>
    <p>
      It's time to get your expenses under control
    </p>
    <button
      className="button"
    >
      Log in with Google
    </button>
  </div>
</div>
`;
